src/backend/base/langflow/services/telemetry/opentelemetry.py

```diff
@@ -1,30 +1,219 @@
+from enum import Enum
 from opentelemetry import metrics
 from opentelemetry.exporter.prometheus import PrometheusMetricReader
+from opentelemetry.metrics import Observation, CallbackOptions
+from opentelemetry.metrics._internal.instrument import Counter, Histogram, UpDownCounter
 from opentelemetry.sdk.metrics import MeterProvider
 from opentelemetry.sdk.resources import Resource
+from typing import Dict, Mapping, Tuple, Union
 
+import threading
+
+# a default OpenTelelmetry meter name
+langflow_meter_name = "langflow"
+
+"""
+If the measurement values are non-additive, use an Asynchronous Gauge.
+    ObservableGauge reports the current absolute value when observed.
+If the measurement values are additive: If the value is monotonically increasing - use an Asynchronous Counter.
+If the value is NOT monotonically increasing - use an Asynchronous UpDownCounter.
+    UpDownCounter reports changes/deltas to the last observed value.
+If the measurement values are additive and you want to observe the distribution of the values - use a Histogram.
+"""
+
+
+class MetricType(Enum):
+    COUNTER = "counter"
+    OBSERVABLE_GAUGE = "observable_gauge"
+    HISTOGRAM = "histogram"
+    UP_DOWN_COUNTER = "up_down_counter"
+
+
+class ObservableGaugeWrapper:
+    """
+    Wrapper class for ObservableGauge
+    Since OpenTelemetry does not provide a way to set the value of an ObservableGauge,
+    instead it uses a callback function to get the value, we need to create a wrapper class.
+    """
+
+    def __init__(self, name: str, description: str, unit: str):
+        self._values: Dict[Tuple[Tuple[str, str], ...], float] = {}
+        self._meter = metrics.get_meter(langflow_meter_name)
+        self._gauge = self._meter.create_observable_gauge(
+            name=name, description=description, unit=unit, callbacks=[self._callback]
+        )
+
+    def _callback(self, options: CallbackOptions):
+        return [Observation(value, attributes=dict(labels)) for labels, value in self._values.items()]
+
+        # return [Observation(self._value)]
+
+    def set_value(self, value: float, labels: Mapping[str, str]):
+        self._values[tuple(sorted(labels.items()))] = value
+
+
+class Metric:
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        type: MetricType,
+        unit: str = "",
+    ):
+        self.name = name
+        self.description = description
+        self.type = type
+        self.unit = unit
+
+    def __repr__(self):
+        return f"Metric(name='{self.name}', description='{self.description}', type={self.type}, unit='{self.unit}')"
+
+
+class ThreadSafeSingletonMeta(type):
+    """
+    Thread-safe Singleton metaclass
+    """
+
+    _instances = None
```

Comment:
I asked chatgpt how singletons are best implemented in python, so feel free to prove it wrong if you know better (I don't). it said using `weakref.WeakValueDictionary()` is preferred, as it allows the garbage collector to remove the singleton instance after deletion.
```
class ThreadSafeSingletonMetaUsingWeakref(type):
    """
    Thread-safe Singleton metaclass using WeakValueDictionary
    """

    _instances = weakref.WeakValueDictionary()
    _lock: threading.Lock = threading.Lock()

    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            with cls._lock:
                if cls not in cls._instances:
                    instance = super(ThreadSafeSingletonMetaUsingWeakref, cls).__call__(*args, **kwargs)
                    cls._instances[cls] = instance
        return cls._instances[cls]
        ```

Reply:
this looks better and I switched to this one. I have a test case to cover race condition, although the previous also passed test.